### ISSUE TITLE
ASGIRef 3.4.0 deadlock_context causes exception on threading

### ISSUE THREAD

**Issue by rdmrocha** [django/asgiref#275]
EDIT:

https://github.com/django/asgiref/issues/275#issuecomment-870083420

**Comment by carltongibson** [comment]
You need to provide more info at the very least. There's nothing to go on here.

**Comment by rdmrocha** [comment]
> You need to provide more info at the very least. There's nothing to go on here.

What type of info do you need? I'm trying to get a boiler-plate which replicates the issue without exposing confidential information but I'm buried in work so I'm taking a bit longer than expected.

Thanks

**Comment by rdmrocha** [comment]
@carltongibson Please check https://github.com/rdmrocha/asgiref-thread-bug

Docker + Django boiler plate, as simple as I could get it.
Clone, and run it. Asgiref triggers the thread exception that generates the eventual 500 errors I first stumbled upon

Edit `requirements.txt` to force the previous version, rerun docker and validate that the error was not there.
This was introduced with https://github.com/django/asgiref/commit/13d0b82a505a753ef116e11b62a6dfcae6a80987

**Comment by carltongibson** [comment]
Thanks for the project @rdmrocha. 

I was going to say this is likely a channels issue but whilst you've got channels installed, you're not actually leveraging it: 

```
from django.core.asgi import get_asgi_application

os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'asgiref_thread_bug.settings')

application = get_asgi_application()
```

That's just Django. 

I'd need to have a dig in to see, but I wonder if the issue is being triggered by the debug Exception Reporter here? 🤔

**Comment by carltongibson** [comment]
@rdmrocha to rule out daphne/channels entirely, the next step is to see if the issue reporoduces using `univorn`.

**Comment by rdmrocha** [comment]
> @rdmrocha to rule out daphne/channels entirely, the next step is to see if the issue reporoduces using `univorn`.

Using uvicorn completely bypasses asgiref

Also, running with `DEBUG=False` exposes the same issue:

![image](https://user-images.githubusercontent.com/3179982/123757748-32a12080-d8b6-11eb-9d15-47e62e371bac.png)

**Comment by rdmrocha** [comment]
This looks like whenever there's an exception (or in my initial scenario, where too much load is on the server), the single thread executer's state gets borked and by raising that `RuntimeError`, it will eventually yield the 500 serving the files.

This example is explicitly forcing an _expected exception_ during the template render. But in the previous scenario, it was just a complex SPA with hundreds of static files being served concurrently and asgiref was randomly tripping on this same raising of the `RuntimeError`, most likely due to a full thread queue or something similar. But these are just wild guesses as I haven't gone deep into ASGIRef's code.

**Comment by raphaelm** [comment]
We have a similar problem. Our situation looks like this:

```
try:
    await sync_to_async(foo)()
except IntegrityError:
    await sync_to_async(bar)()
```

This fails with the same RuntimeError if the ``except``-case is triggered.

**Comment by carltongibson** [comment]
Maybe related, a report of a runworker issue coming up with asgiref 3.4 (and working before)

https://github.com/django/channels/issues/1713

(I'm hoping for more info on this last.)

**Comment by AMontagu** [comment]
Same problem when using grpc.aio with python on big load of data.
Putting thread_sensitive=False remove the problem. 

Is there any possible bug with thread_sensitive=False and using the django ORM like:
`project = await sync_to_async(serializer.save, thread_sensitive=False)()`

I am relatively new in async in python and have some difficulty to understand what going on when using sync_to_async when performing action in db.

**Comment by andrewgodwin** [comment]
I was worried we might find an edge case like this - I suspect some of these are genuine deadlock prevention, and some are unfortunate side-effects of the detection code. It's very hard to navigate all the possible cases here.

If you can set it, `thread_sensitive=False` will of course disable the detection because you're opting out of the safe mode, but make sure your code doesn't contain any shared database connections, file handles or similar. Otherwise, stick to an earlier version while we work this out.

The `except` handling case does seem suspicious to me - this might be why it's happening in the original case, too. I'll have to sit down and work through it later this week.

**Comment by andrewgodwin** [comment]
Turns out this was a bit easier than expected - we were just missing a `finally`: https://github.com/django/asgiref/commit/e8797d47f4006e4e817fda507ec1eacfed4faf60

If someone here can test and verify it fixes their issue, I will cut a release later tonight.

**Comment by raphaelm** [comment]
Thank you! <3

**Comment by andrewgodwin** [comment]
This fix is now released in `3.4.1`.
